thejcpalma/palworld-dedicated-server-docker
# Comprehensive Update for Project Improvements and Palworld Update 0.1.5 Compatibility

# Comprehensive Update for Project Improvements and Palworld Update `0.1.5` Compatibility

## Summary

This pull request includes a series of changes aimed at improving the project structure, enhancing security, updating documentation, implementing player activity monitoring, and addressing the breaking changes introduced by the `0.1.5` Palworld update.

## Motivation and Context

The motivation behind these changes is to make it easier for others to contribute to the project, improve player activity monitoring, and ensure the server's compatibility with the latest version of Palworld. Additionally, minor fixes have been made to enhance the security and ease of sharing logs.

Improving the project structure and documentation to make it easier for others to contribute to the project.
Implementing a player activity monitoring that logs player join and leave events on both the logs and the discord webhook.
Minor fixes to make sharing logs easier, like redacting password and the public IP.
Ensure that `0.1.5` Palworld update works. It introduced breaking changes that affected the functionality of our server. This ensures our server continues to function correctly with the latest version of Palworld.
New configuration logic using new utility function, template files and environment variables substitution to improve future changes.

## Description

The changes in this pull request include:

- Adding a custom RCON binary for broadcast command to be able to send spaces
- Implementing player activity monitoring that logs player join and leave events on both the logs and the discord webhook
- Redacting passwords and public IP in the logs for safe sharing
- Restructuring the project files to be more atomic, making it easier to implement new features and collaborate on the project
- Updating the webhook documentation to include player monitoring and emoji usage in titles and descriptions
- Adding a new case for webhook messages when the server is restarting and the time is 0 or no players are online
- Fixing the breaking changes introduced by the `0.1.5` Palworld update
- Refactor configuration setup:
  - Improved file structure to be more atomic and more concise
  - Implemented new utility functions for easier configuration handling
- Add Templating to All Configuration Files:
  - Added [`PalWorldSetting.ini.template`](https://github.com/thejcpalma/palworld-dedicated-server-docker/blob/0.1.4/scripts/config/templates/PalWorldSettings.ini.template) for dynamic configuration
  - Added  [`Engine.ini.template`](https://github.com/thejcpalma/palworld-dedicated-server-docker/blob/0.1.4/scripts/config/templates/Engine.ini.template) for dynamic configuration
  - Re-added rcon.yaml (as [`rcon.yaml.template`](https://github.com/thejcpalma/palworld-dedicated-server-docker/blob/0.1.4/scripts/config/templates/rcon.yaml.template)) to use the new configuration logic
- Fix log functions expanding escape sequences edge cases:
  - Addressed specific edge cases where log functions were expanding escape sequences incorrectly.
  - Improved the robustness of log functions to handle various scenarios.
- Add possibility for special characters in the player name for webhooks:
  - Enhanced player name handling in webhooks to support special characters.
  - Improved compatibility and user experience for diverse player names.
- Add `jq` package to handle Discord Webhook messages:
  - Introduced the `jq` package to streamline the processing of Discord Webhook messages.
  - Enhances readability and manipulation of JSON data within the webhook messages.
- Updating the documentation to reflect the changes made in this pull request

## Testing Instructions

These changes were tested by running the server with the latest version of Palworld and verifying that it functions correctly:
 - Build image correctly and without errors
 - Run the server with the new image
 - Verify that the server starts correctly
 - Verify that all the previous features work correctly
 - Verify that the server logs player join and leave events
 - Verify that the server logs the correct information to the discord webhook
 - Verify that the server redacts passwords and public IP in the logs
 - Verify that the server can send spaces in the broadcast command

Names tested on the webhook events to check if both are logged correctly on docker logs and discord webhook:
```yaml

Oiginal |p* \n y,er\t \v,,,""
LogInfo |p* \n y,er\t \v,,,""
Discord |p* \n y,er\t \v,,,""

Oiginal |   ??test{[(,.-_|\e 
LogInfo |   ??test{[(,.-_|\e 
Discord |   ??test{[(,.-_|\e 

Oiginal |\$$\\n'"te,st\ \x '
LogInfo |\$$\\n'"te,st\ \x '
Discord |\$$\\n'"te,st\ \x '


Oiginal |*t,et*** \* \? $ \$
LogInfo |*t,et*** \* \? $ \$
Discord |*t,et*** \* \? $ \$


Oiginal |__\$$$\\** \*__ $@
LogInfo |__\$$$\\** \*__ $@
Discord |__\$$$\\** \*__ $@

Oiginal |``_   \\\\
LogInfo |``_   \\\\
Discord |''_   \\\\

Oiginal |``_   \\$$$$$$\\||__
LogInfo |``_   \\$$$$$$\\||__
Discord |''_   \\$$$$$$\\||__

Oiginal |``_   \\$$$\\||__ ``
LogInfo |``_   \\$$$\\||__ ``
Discord |''_   \\$$$\\||__ ''

Oiginal |@everyone\\$$$test
LogInfo |@everyone\\$$$test
Discord |@everyone\\$$$test

Backticks ( ` ) are always replaced with single quotes ( ' ) on the names when
sending to the discord webhook to prevent format harrassment.
```

The updated documentation was reviewed for clarity and accuracy.

## Checklist
